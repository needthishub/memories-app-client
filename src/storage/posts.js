import * as api from '../api';

const CREATE = 'POSTS/CREATE';
const UPDATE = 'POSTS/UPDATE';
const DELETE = 'POSTS/DELETE';
const FETCH_ALL = 'POSTS/FETCH_ALL';
const LIKE = 'POSTS/LIKE';
const ERROR = 'POSTS/ERROR';

export const fetchPostsAction = () => async (dispatch) => {
  try {
    const { data } = await api.fetchPosts();

    dispatch({ type: FETCH_ALL, payload: data });
  } catch (error) {
    console.log(error.message);
    dispatch({ type: ERROR, payload: error.message });
  }
};

export const createPostAction = (post) => async (dispatch) => {
  try {
    const { data } = await api.createPost(post);

    dispatch({ type: CREATE, payload: data });
  } catch (error) {
    console.log(error.message);
    dispatch({ type: ERROR, payload: error.message });
  }
};
export const updatePostAction = (id, post) => async (dispatch) => {
  try {
    const { data } = await api.updatePost(id, post);

    dispatch({ type: UPDATE, payload: data });
  } catch (error) {
    console.log(error.message);
    dispatch({ type: ERROR, payload: error.message });
  }
};
export const likePostAction = (id) => async (dispatch) => {
  try {
    const { data } = await api.likePost(id);

    dispatch({ type: LIKE, payload: data });
  } catch (error) {
    console.log(error.message);
    dispatch({ type: ERROR, payload: error.message });
  }
};
export const deletePostAction = (id) => async (dispatch) => {
  try {
    await api.deletePost(id);

    dispatch({ type: DELETE, payload: id });
  } catch (error) {
    console.log(error.message);
    dispatch({ type: ERROR, payload: error.message });
  }
};

export const selectPostById = (currentId) => (state) => (
  currentId ? state.posts.list.find((message) => message._id === currentId) : null);

export const selectPosts = (state) => state.posts.list;

export const selectError = (state) => state.posts.error;

const INITIAL_STATE = { list: [], error: null };

export default (state = INITIAL_STATE, { type, payload }) => {
  switch (type) {
    case FETCH_ALL:
      return { list: payload, error: null };
    case LIKE:
      return { list: state.list.map((post) => (post._id === payload._id ? payload : post)), error: null };
    case CREATE:
      return { list: [...state.list, payload], error: null };
    case UPDATE:
      return { list: state.list.map((post) => (post._id === payload._id ? payload : post)), error: null };
    case DELETE:
      return { list: state.list.filter((post) => post._id !== payload), error: null };
    case ERROR:
      return { ...state, error: payload };
    default:
      return state;
  }
};
