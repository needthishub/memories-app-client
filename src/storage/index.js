import { applyMiddleware, combineReducers, compose, createStore } from 'redux';

import thunk from 'redux-thunk';
import posts from './posts';

const reducers = combineReducers({ posts });
export const createStorage = (initialState = {}) => (
  createStore(reducers, initialState, compose(applyMiddleware(thunk))));

