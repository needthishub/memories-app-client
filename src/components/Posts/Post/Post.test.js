import React from 'react';
import { shallow } from 'enzyme';
import * as reactRedux from 'react-redux';
import Post from './Post';
import { deletePostAction, likePostAction } from '../../../storage/posts';

jest.mock('../../../storage/posts');

describe('<Post />', () => {
  const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');
  const testID = 1;

  beforeEach(() => {
    useDispatchMock.mockReturnValue(jest.fn());
  });

  afterEach(() => {
    useDispatchMock.mockClear();
  });

  it('should set post id while click on button', () => {
    const currentIdMock = jest.fn();
    const wrapper = shallow(<Post
      post={{
        selectedFile: 'testImage',
        title: 'testTitle',
        createdAt: jest.fn(),
        _id: testID,
        tags: ['testTag1', 'testTag2', 'testTag3', 'testTag4', 'testTag5'],
        message: 'testMessage',
        likeCount: '1' }}
      setCurrentId={currentIdMock}
    />);
    wrapper.dive().find('.edit-btn').props().onClick();
    expect(currentIdMock).toBeCalledWith(testID);
  });

  it('should like post while click on button', () => {
    const wrapper = shallow(<Post
      post={{
        selectedFile: 'testImage',
        title: 'testTitle',
        createdAt: jest.fn(),
        _id: testID,
        tags: ['testTag1', 'testTag2', 'testTag3', 'testTag4', 'testTag5'],
        message: 'testMessage',
        likeCount: '1' }}
      setCurrentId={jest.fn()}
    />);
    wrapper.find('.like-btn').props().onClick();
    expect(likePostAction).toBeCalledWith(testID);
  });

  it('should delete post while click on button', () => {
    const wrapper = shallow(<Post
      post={{
        selectedFile: 'testImage',
        title: 'testTitle',
        createdAt: jest.fn(),
        _id: testID,
        tags: ['testTag1', 'testTag2', 'testTag3', 'testTag4', 'testTag5'],
        message: 'testMessage',
        likeCount: '1' }}
      setCurrentId={jest.fn()}
    />);
    wrapper.find('.delete-btn').props().onClick();
    expect(deletePostAction).toBeCalledWith(testID);
  });

  it('should match to snapshot', () => {
    const wrapper = shallow(<Post
      post={{
        selectedFile: 'testImage',
        title: 'testTitle',
        createdAt: jest.fn(),
        _id: testID,
        tags: ['testTag1', 'testTag2', 'testTag3', 'testTag4', 'testTag5'],
        message: 'testMessage',
        likeCount: '1' }}
      setCurrentId={jest.fn()}
    />);
    expect(wrapper).toMatchSnapshot();
  });
});
