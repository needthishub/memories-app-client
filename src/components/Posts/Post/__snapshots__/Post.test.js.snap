// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Post /> should match to snapshot 1`] = `ShallowWrapper {}`;
